### PR TITLE
Support for multiple X509 certificates

### PR DESCRIPTION
**Pull request by davelooi**
# Description

Official ruby-saml does not handle multiple X509 certificates in metadata. This feature is required for rotating certificate provided by SAML provider like Microsoft Azure.

This PR does:
- extracts all X509 certs from metadata
- stores certs and fingerprints in array
- validates response against array of fingerprints or certs
# Approval
- [x] lead @sentience
- [x] non-lead @bjz 
# Post merge

Tag the special release.

### CODE REVIEW

**Comment by davelooi** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;L7HQgvsy&#x2F;154-fork-ruby-saml-gem-and-make-it-support-multiple-certificates"> Fork ruby-saml gem and make it support multiple certificates</a>

**Comment by sentience** [comment]
Looks good. Just recommend naming all your “multi-cert” versions of methods `_multicert` instead of just pluralising them with an `s` on the end. That’ll make it clearer why they exist, and is something that’s easy to grep for.

File: README.md [hunk 1]
```diff
@@ -1,588 +1,17 @@
...
 
-settings.attributes_index = 5
-settings.attribute_consuming_service.configure do
-  service_name "Service"
-  service_index 5
-  add_attribute :name => "Name", :name_format => "Name Format", :friendly_name => "Friendly Name"
-  add_attribute :name => "Another Attribute", :name_format => "Name Format", :friendly_name => "Friendly Name", :attribute_value => "Attribute Value"
-end
-```
+- IdpMetadataParser will parse ALL X509 certificates in the metadata
```
Comment (brendanzab) [inline README.md]: Nit - backtickify `IdpMetadataParser`

File: lib/onelogin/ruby-saml/idp_metadata_parser.rb [hunk 2]
```diff
@@ -147,6 +150,22 @@ def certificate_base64
         end
       end
 
+      # @return Array of Unformatted Certificates
+      #
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/idp_metadata_parser.rb:154]: Looks like they are using [YARD](http://yardoc.org/types.html). Perhaps:

``` ruby
# @return [Array<String>] Unformatted Certificates
```

File: lib/onelogin/ruby-saml/idp_metadata_parser.rb [hunk 3]
```diff
@@ -155,6 +174,17 @@ def certificate
         end
       end
 
+      # @return Array of X509Certificates
+      #
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/idp_metadata_parser.rb:178]: Ditto above:

``` ruby
# @return [Array<String>] X509Certificates
```

File: lib/onelogin/ruby-saml/idp_metadata_parser.rb [hunk 4]
```diff
@@ -166,6 +196,19 @@ def fingerprint
           end
         end
       end
+
+      # @return Array of the SHA-1 fingerpints of the X509Certificates
+      #
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/idp_metadata_parser.rb:199]: Ditto above:

``` ruby
# @return [Array<String>] the SHA-1 fingerpints of the X509Certificates
```

File: lib/onelogin/ruby-saml/idp_metadata_parser.rb [hunk 5]
```diff
@@ -155,6 +174,17 @@ def certificate
...
 
+      # @return Array of X509Certificates
+      #
+      def certificate_multi
+        @certificate_multi ||= begin
+          certs = []
+          certificate_base64_multi.each do |certificate_base64|
+            certs << Base64.decode64(certificate_base64)
+          end
+          certs
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/idp_metadata_parser.rb]:
``` ruby
certificate_base64_multi.map do |certificate_base64|
  Base64.decode64(certificate_base64)
end
```

File: lib/onelogin/ruby-saml/idp_metadata_parser.rb [hunk 6]
```diff
@@ -166,6 +196,19 @@ def fingerprint
...
+      # @return Array of the SHA-1 fingerpints of the X509Certificates
+      #
+      def fingerprint_multi
+        @fingerprint_multi ||= begin
+          fingerprints = []
+          certificate_multi.each do |certificate|
+            cert = OpenSSL::X509::Certificate.new(certificate)
+            fingerprints << Digest::SHA1.hexdigest(cert.to_der).upcase.scan(/../).join(":")
+          end
+          fingerprints
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/idp_metadata_parser.rb]: Could be a `map`

File: lib/onelogin/ruby-saml/settings.rb [hunk 7]
```diff
@@ -113,6 +115,23 @@ def get_fingerprint
         end
       end
 
+      # Calculates the fingerprints of the IdP x509 certificates.
+      # @return [Array of String] The fingerprints
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/settings.rb]: `Array<String>`

File: lib/onelogin/ruby-saml/settings.rb [hunk 8]
```diff
@@ -122,6 +141,19 @@ def get_idp_cert
         OpenSSL::X509::Certificate.new(formatted_cert)
       end
 
+      # @return [Array of OpenSSL::X509::Certificate] Build the IdP certificates from the settings (previously format it)
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/settings.rb]: `Array<OpenSSL::X509::Certificate>`

File: test/test_helper.rb [hunk 9]
```diff
@@ -183,14 +191,40 @@ def ruby_saml_cert
...
   end
 
+  def ruby_saml_certs
+    @ruby_saml_certs ||= begin
+      certs = []
+      ruby_saml_certs_text.each do |ruby_saml_cert_text|
+        certs << OpenSSL::X509::Certificate.new(ruby_saml_cert_text)
+      end
+      certs
+    end
```
Comment (brendanzab) [inline test/test_helper.rb:199]: Coud map this too, and the other one below.

**Comment by brendanzab** [comment]
Logic looks fine to me, but I'm not a SAML expert. There are a bunch of places where you could use `map`s instead of exposed accumulations, but that is just my preference. Would be good to use the proper type syntax for the doc comments too - would reduce the friction for an eventual(?) merge upstream.

File: lib/onelogin/ruby-saml/response.rb [hunk 10]
```diff
@@ -596,6 +629,40 @@ def validate_signature
...
+          "/p:Response[@ID=$id]",
+          { "p" => PROTOCOL, "ds" => DSIG },
+          { 'id' => document.signed_element_id }
+        )
+        doc = (response_signed || decrypted_document.nil?) ? document : decrypted_document
+
+        fingerprint_multi.zip(idp_cert_multi).any? do |fingerprint, idp_cert|
+          opts = {}
+          opts[:fingerprint_alg] = settings.idp_cert_fingerprint_algorithm
+          opts[:cert] = idp_cert
```
Comment (brendanzab) [inline lib/onelogin/ruby-saml/response.rb:660]: This is kind of ugly, but I agree it's probably better we stick close to the logic in `#validate_signature`

**Comment by brendanzab** [comment]
:+1: :ship: :shipit:

File: lib/onelogin/ruby-saml/response.rb (hunk 10 above)
Comment (brendanzab) [inline lib/onelogin/ruby-saml/response.rb:660]: Just clarifying - talking about the `opts` not using a literal.

**Pull request merged**